Спринт 14

il76 opened the pull request.

Commit:
Прогрессбар

Commit:
Дебаунсер для кликов по тексту.
Экспериментируем с ViewBinding.
Частично дебаунсер для поиска, но пока нерабочий.

Commit:
Дебаунсер для поиска.
Клик по кнопке "поиск" оставил, при этом сбрасываем очередь задач, чтобы запрос повторно не дублировался.

Commit:
SDK 35
Экран плеера переделал на вьюбиндинг

Commit:
Переключатель состояния кнопок

Commit:
Рабочая версия плеера, без отслеживания прогресса воспроизведения

Commit:
Прогресс воспроизведения

Commit:
Поменял прейслолдер на ресурсы и пока раскомментировал перестроение истории при клике.

kekulta commented on app/src/main/java/com/il76/playlistmaker/PlayerActivity.kt:
```diff
@@ -33,40 +67,161 @@ class PlayerActivity : AppCompatActivity() {
-        val tvCountry = findViewById<TextView>(R.id.trackCountry)
-        tvCountry.text = track.country
+    companion object {
+        private const val STATE_DEFAULT = 0
+        private const val STATE_PREPARED = 1
+        private const val STATE_PLAYING = 2
+        private const val STATE_PAUSED = 3
```
Стейт удобнее моделировать с помощью `enum` или `sealed interface`.

kekulta commented on app/src/main/java/com/il76/playlistmaker/SearchActivity.kt:
```diff
@@ -180,20 +220,29 @@ class SearchActivity : AppCompatActivity() {
+                                trackList.addAll(trackSearchHistory.trackListHistory.reversed())
+                                trackAdapter.notifyDataSetChanged()
+                            }
+                            val json = Gson().toJson(elem)
```
Создание `Gson()` это дорогая операция. В идеале, мы хотим создавать его один раз за время жизни приложения и переиспользовать.

kekulta commented on app/src/main/java/com/il76/playlistmaker/SearchActivity.kt:
```diff
@@ -142,13 +186,14 @@ class SearchActivity : AppCompatActivity() {
         val inputEditText = findViewById<EditText>(R.id.search_edit_text)
         inputEditText.setOnEditorActionListener { _, actionId, _ ->
             if (actionId == EditorInfo.IME_ACTION_DONE) {
+                handler.removeCallbacks(searchRunnable)
                 doSearch()
```
Молодец, что сразу поймал этот корнер-кейс!

kekulta commented on app/src/main/java/com/il76/playlistmaker/SearchActivity.kt:
```diff
@@ -142,13 +186,14 @@ class SearchActivity : AppCompatActivity() {
         }
 
-        val clearButton = findViewById<ImageView>(R.id.search_icon_clear)
+        clearButton = findViewById<ImageView>(R.id.search_icon_clear)
```
Вижу, что выше включил байндинг, непонятно откуда тут взялся findView.

kekulta commented on app/src/main/java/com/il76/playlistmaker/SearchActivity.kt:
```diff
@@ -142,13 +186,14 @@ class SearchActivity : AppCompatActivity() {
+        clearButton = findViewById<ImageView>(R.id.search_icon_clear)
         clearButton.setOnClickListener {
             inputEditText.setText("")
             val inputMethodManager = getSystemService(Context.INPUT_METHOD_SERVICE) as? InputMethodManager
```
`getSystemService<InputMethodManager>()`

Commit:
Gson создаём при запуске приложения.
Большую часть findViewById переписал на работу с биндингом.
Единая функция логирования для упрощения отладки.
